Implement efficient size method for Tries

gregw opened the pull request:
The isEmpty method can be frequently called on tries and the current implementation creates a disposible key set each time.

gregw committed:
Implement efficient size method for Tries

The isEmpty method can be frequently called on tries and the current implementation creates a disposible key set each time.

gregw commented:
Seen in #13622 - whilst not the cause, the depth was not good either.
It may also be worthwhile looking to avoid tail recursion when possible.,

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:
```diff
@@ -137,6 +138,7 @@ public String toString()
      *
      * The array is of characters rather than integers to save space.
      */
+    private final AtomicInteger _size = new AtomicInteger();
```
`ArrayTrie` isn't thread-safe, so that atomic can be replaced by a simple int.

sbordet replied:
Why atomic? It's supposed to be non-concurrent.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java:
```diff
@@ -97,6 +98,7 @@ private Node(char c)
         }
     }
 
+    private final AtomicInteger _size = new AtomicInteger();
```
`TreeTrie` also isn't thread-safe, so int can be used here too.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/ArrayTrie.java:
```diff
@@ -296,6 +299,10 @@ else if (idx >= big.length)
         node._key = key;
+        if (node._value == null && value != null)
+            _size.incrementAndGet();
+        else if (node._value != null && value == null)
```
The javadocs of the `put()` method say that `value` cannot be `null`, but `remove()` obviously breaks that.

Furthermore, what's the point of creating the entry, just to have a `null` value at the end?

gregw replied:
This PR is just adding size.  I don't want to change the way the Trie works.  If a node was created before it is still created after this PR

gregw replied:
Also we actually have tests that check we can put null values, so I have changed the javadoc

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java:
```diff
@@ -156,8 +159,20 @@ public boolean put(String s, V v)
         }
-        t._key = v == null ? null : s;
-        t._value = v;
+        if (v == null)
```
Same comment... This is creating all the nodes.

gregw replied:
same response.  This PR is not changing the way the Trie works, just adding a size

gregw committed:
Implement efficient size method for Tries

The isEmpty method can be frequently called on tries and the current implementation creates a disposible key set each time.

gregw committed:
remove accidental commit

sbordet commented on jetty-core/jetty-server/src/test/resources/jetty-logging.properties:
```diff
@@ -1,4 +1,4 @@
-#org.eclipse.jetty.LEVEL=DEBUG
+Doorg.eclipse.jetty.LEVEL=DEBUG
```
Please revert.

sbordet commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/TreeTrie.java:
```diff
@@ -276,37 +290,13 @@ private V getBest(Node<V> node, byte[] b, int offset, int len)
-            }
-        }
-        return true;
+        return _size == 0;
```
```suggestion
        return size() == 0;
```

gregw committed:
updates from review
